### PR TITLE
Modifications to support Python 3.

### PR DESCRIPTION
**Pull request by hydratk**
The modifications were made on tag 1.6.0 (which is available on PyPi).
I updated stompest.core, stompest.async is kept untouched (I don't use Twisted)
Unit tests passed on Python 2.6, 2.7, 3.3, 3.4, 3.5

### CODE REVIEW

File: src/core/stompest/protocol/commands.py [hunk 1]
```diff
@@ -120,7 +121,10 @@ def subscribe(destination, headers, receipt=None, version=None):
         if (version != StompSpec.VERSION_1_0):
             raise
     token = (StompSpec.DESTINATION_HEADER, destination) if (subscription is None) else (StompSpec.ID_HEADER, subscription)
-    return frame, tuple(map(unicode, token))
```
Comment (nikipore) [inline src/core/stompest/protocol/commands.py]: Please factor out the distinction of cases s.t. no redundant code remains, e.g. using a ternary `... if ... else ...`).

File: src/core/stompest/protocol/frame.py [hunk 2]
```diff
@@ -82,9 +83,13 @@ def __repr__(self):
 
     def __str__(self):
         """Render the wire-level representation of a STOMP frame."""
-        headers = sorted(self.headers.iteritems()) if self.rawHeaders is None else self.rawHeaders
-        headers = ''.join('%s:%s%s' % (self._encode(self._escape(unicode(key))), self._encode(self._escape(unicode(value))), StompSpec.LINE_DELIMITER) for (key, value) in headers)
-        return StompSpec.LINE_DELIMITER.join([self._encode(unicode(self.command)), headers, '%s%s' % (self.body, StompSpec.FRAME_DELIMITER)])
+        headers = sorted(self.headers.items()) if self.rawHeaders is None else self.rawHeaders
```
Comment (nikipore) [inline src/core/stompest/protocol/frame.py]: Please factor out the distinction of cases s.t. no redundant code remains, e.g. `cast = unicode if isPython3() else str`).

File: src/core/stompest/protocol/frame.py [hunk 3]
```diff
@@ -106,7 +111,7 @@ def version(self, value):
         self._version = StompSpec.version(value)
 
     def _encode(self, text):
-        return StompSpec.CODECS[self.version].encode(text)[0]
```
Comment (nikipore) [inline src/core/stompest/protocol/frame.py]: Please factor out the distinction of cases s.t. no redundant code remains, e.g. by appending an `if` statement which decodes the result.

File: src/core/stompest/protocol/parser.py [hunk 4]
```diff
@@ -64,6 +69,8 @@ def add(self, data):
         
         :param data: An iterable of characters. If any character evaluates to :obj:`False`, that stream will no longer be consumed.
         """
+        if version_info[0] == 3 and data.__class__.__name__ not in ['str', 'generator']:
```
Comment (nikipore) [inline src/core/stompest/protocol/parser.py]: Don't refer to class names. You should use either duck-typing or `isinstance` or some other polymorphic mechanism.

File: src/core/stompest/protocol/parser.py [hunk 5]
```diff
@@ -95,6 +102,8 @@ def _transition(self, state):
         self.parse = self._parsers[state]
 
     def _parseHeartBeat(self, character):
+        if version_info[0] == 3 and character.__class__.__name__ != 'str':
```
Comment (nikipore) [inline src/core/stompest/protocol/parser.py]: see above

File: src/core/stompest/protocol/parser.py [hunk 6]
```diff
@@ -105,6 +114,8 @@ def _parseHeartBeat(self, character):
             self._append()
 
     def _parseCommand(self, character):
+        if version_info[0] == 3 and character.__class__.__name__ != 'str':
```
Comment (nikipore) [inline src/core/stompest/protocol/parser.py]: see above

File: src/core/stompest/protocol/parser.py [hunk 7]
```diff
@@ -115,6 +126,8 @@ def _parseCommand(self, character):
         self._transition('headers')
 
     def _parseHeader(self, character):
+        if version_info[0] == 3 and character.__class__.__name__ != 'str':
```
Comment (nikipore) [inline src/core/stompest/protocol/parser.py]: see above

File: src/core/stompest/protocol/parser.py [hunk 8]
```diff
@@ -132,6 +145,8 @@ def _parseHeader(self, character):
             self._transition('body')
 
     def _parseBody(self, character):
+        if version_info[0] == 3 and character.__class__.__name__ != 'str':
```
Comment (nikipore) [inline src/core/stompest/protocol/parser.py]: see above

File: src/core/stompest/protocol/spec.py [hunk 9]
```diff
@@ -62,7 +62,7 @@ class StompSpec(object):
         VERSION_1_0: 'ascii'
     }
     CODECS = dict([
-        (version, codecs.lookup(CODECS.get(version, 'utf-8'))) for version in VERSIONS
+        (version, codecs.lookup('ascii' if version == '1.0' else 'utf-8')) for version in VERSIONS
```
Comment (nikipore) [inline src/core/stompest/protocol/spec.py]: I don't understand why you changed this. It's ok, but please remove ll. 61-63.

File: src/core/stompest/sync/client.py [hunk 10]
```diff
@@ -289,11 +289,11 @@ def canRead(self, timeout=None):
             timeout = deadline and max(0, deadline - time.time())
             if not self._transport.canRead(timeout):
                 return False
-            frame = self._transport.receive()
+            frame = self._transport.receive()
```
Comment (nikipore) [inline src/core/stompest/sync/client.py]: remove trailing whitespace

File: src/core/stompest/sync/client.py [hunk 11]
```diff
@@ -289,11 +289,11 @@ def canRead(self, timeout=None):
             timeout = deadline and max(0, deadline - time.time())
             if not self._transport.canRead(timeout):
                 return False
-            frame = self._transport.receive()
+            frame = self._transport.receive()            
             self.session.received()
             if self.log.isEnabledFor(logging.DEBUG):
-                self.log.debug('Received %s' % frame.info())
-            if frame:  # there's a real STOMP frame on the wire, not a heart-beat
+                self.log.debug('Received %s' % frame.info())
```
Comment (nikipore) [inline src/core/stompest/sync/client.py]: remove trailing whitespace

File: src/core/stompest/sync/client.py [hunk 12]
```diff
@@ -289,11 +289,11 @@ def canRead(self, timeout=None):
...
             if not self._transport.canRead(timeout):
                 return False
-            frame = self._transport.receive()
+            frame = self._transport.receive()            
             self.session.received()
             if self.log.isEnabledFor(logging.DEBUG):
-                self.log.debug('Received %s' % frame.info())
-            if frame:  # there's a real STOMP frame on the wire, not a heart-beat
+                self.log.debug('Received %s' % frame.info())            
+            if frame.__class__.__name__ == 'StompFrame':  # there's a real STOMP frame on the wire, not a heart-beat
```
Comment (nikipore) [inline src/core/stompest/sync/client.py]: Why did you remove the duck-typing? Explain or revert!

File: src/core/stompest/sync/transport.py [hunk 13]
```diff
@@ -72,7 +73,11 @@ def receive(self):
             self._parser.add(data)
 
     def send(self, frame):
-        self._write(str(frame))
+        if version_info[0] == 3:
```
Comment (nikipore) [inline src/core/stompest/sync/transport.py]: Extract `self._write(frame)` from the `if` clause.

File: src/core/stompest/util/__init__.py [hunk 14]
```diff
@@ -22,6 +23,6 @@ def cloneFrame(frame, persistent=None):
     frame.unraw()
     headers = filterReservedHeaders(frame.headers)
     if persistent is not None:
-        headers[u'persistent'] = unicode(bool(persistent)).lower()
+        headers[u'persistent'] = unicode(bool(persistent)).lower() if version_info[0] == 2 else str(bool(persistent)).lower()
```
Comment (nikipore) [inline src/core/stompest/util/__init__.py]: Extract `cast = unicode if isPython3() else str`. Probably `cast`and `isPython2()` and `cast` should be `util` functions.

File: src/core/stompest/protocol/commands.py (hunk 1 above)
Comment (hydratk) [inline src/core/stompest/protocol/commands.py]: updated

File: src/core/stompest/protocol/frame.py (hunk 2 above)
Comment (hydratk) [inline src/core/stompest/protocol/frame.py]: updated

File: src/core/stompest/protocol/frame.py (hunk 3 above)
Comment (hydratk) [inline src/core/stompest/protocol/frame.py]: updated

File: src/core/stompest/protocol/parser.py (hunk 4 above)
Comment (hydratk) [inline src/core/stompest/protocol/parser.py]: updated

File: src/core/stompest/protocol/parser.py (hunk 5 above)
Comment (hydratk) [inline src/core/stompest/protocol/parser.py]: updated

File: src/core/stompest/protocol/parser.py (hunk 6 above)
Comment (hydratk) [inline src/core/stompest/protocol/parser.py]: updated

File: src/core/stompest/protocol/parser.py (hunk 7 above)
Comment (hydratk) [inline src/core/stompest/protocol/parser.py]: updated

File: src/core/stompest/protocol/parser.py (hunk 8 above)
Comment (hydratk) [inline src/core/stompest/protocol/parser.py]: updated

File: src/core/stompest/protocol/spec.py (hunk 9 above)
Comment (hydratk) [inline src/core/stompest/protocol/spec.py]: CODECS dictionary was not visible in Py3 interpreter.

File: src/core/stompest/sync/client.py (hunk 10 above)
Comment (hydratk) [inline src/core/stompest/sync/client.py]: updated

File: src/core/stompest/sync/client.py (hunk 11 above)
Comment (hydratk) [inline src/core/stompest/sync/client.py]: updated

File: src/core/stompest/sync/client.py (hunk 12 above)
Comment (hydratk) [inline src/core/stompest/sync/client.py]: In Py3 interpreter StompHeartBeat frames passed through the check and caused problems, so I changed to check to be more explicit.

File: src/core/stompest/sync/transport.py (hunk 13 above)
Comment (hydratk) [inline src/core/stompest/sync/transport.py]: updated

File: src/core/stompest/util/__init__.py (hunk 14 above)
Comment (hydratk) [inline src/core/stompest/util/__init__.py]: updated, module stompest.protocol.util provides function ispy2
